### ISSUE TITLE
\yii\db\ActiveQuery::count() ignores named expressions

### ISSUE THREAD

**Issue by herroffizier** [yiisoft/yii2#8068]
I have some named expression declared in ```SELECT``` part of query:
```php
$query = \app\models\Provider::find()->
            select(['id', 'val' => new \yii\db\Expression('(id * 2)')]);
```
In real case I use a little bit more sensible expression, of course, but it doesn't matter for now. Next, I want to filter results by declared expression's value:
```php
$query->having('val > 2');
```
Now, if I would call ```$query->all()```, correct results would be returned, but if I would call ```count()``` I would get following error:
```
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'val' in 'having clause'
The SQL being executed was: SELECT COUNT(*) FROM `provider` HAVING val > 2
```
As you may notice, generated query is ```SELECT COUNT(*) FROM `provider` HAVING val > 2``` while, as I assume, it should be something like ```SELECT COUNT(*) FROM (SELECT (id * 2) AS val FROM `provider` HAVING val > 2) c``` but such query is genererated only if ```groupBy``` or ```union``` is not empty or query has ```distinct = true```.

Is it a bug or I missed something?

I am using Yii 2.0.3 on PHP 5.6.7 with MySQL database.

**Comment by cebe** [comment]
the current implementation does not cover this use case, could be improved.

**Comment by klimov-paul** [comment]
Check for 'having' present should be added here:
https://github.com/yiisoft/yii2/blob/master/framework/db/Query.php#L388

**Comment by klimov-paul** [comment]
Well, not I am not so sure. If you run a query with `HAVING` but without group by (which presence is covered), you will get 'General error: 1 a GROUP BY clause is required before HAVING'.

**Comment by klimov-paul** [comment]
No, MySQL allows it.

**Comment by klimov-paul** [comment]
Issue resolved by commit b8081b5

**Comment by cebe** [comment]
:+1:
